pkg/admissioncontroller/webhook/admission/resourcesize/handler.go

```diff
@@ -31,94 +34,78 @@ import (
 	apisconfig "github.com/gardener/gardener/pkg/admissioncontroller/apis/config"
 	confighelper "github.com/gardener/gardener/pkg/admissioncontroller/apis/config/helper"
 	"github.com/gardener/gardener/pkg/admissioncontroller/metrics"
-	acadmission "github.com/gardener/gardener/pkg/admissioncontroller/webhook/admission"
-	"github.com/gardener/gardener/pkg/logger"
 )
 
-const (
-	// HandlerName is the name of this admission webhook handler.
-	HandlerName = "resource_size_validator"
-	// WebhookPath is the HTTP handler path for this admission webhook handler.
-	WebhookPath = "/webhooks/validate-resource-size"
+// metricReasonSizeExceeded is a metric reason value for a reason when an object size was exceeded.
+const metricReasonSizeExceeded = "Size Exceeded"
 
-	// metricReasonSizeExceeded is a metric reason value for a reason when an object size was exceeded.
-	metricReasonSizeExceeded = "Size Exceeded"
-)
-
-// New creates a new webhook handler validating that the resource size of a request doesn't exceed the configured
-// limits.
-func New(logger logr.Logger, config *apisconfig.ResourceAdmissionConfiguration) *handler {
-	return &handler{logger: logger, config: config}
-}
-
-type handler struct {
-	logger  logr.Logger
-	config  *apisconfig.ResourceAdmissionConfiguration
-	decoder *admission.Decoder
-}
-
-var _ admission.Handler = &handler{}
-
-func (h *handler) InjectDecoder(d *admission.Decoder) error {
-	h.decoder = d
-	return nil
+// Handler checks the resource sizes.
+type Handler struct {
+	Logger logr.Logger
+	Config *apisconfig.ResourceAdmissionConfiguration
 }
 
-func (h *handler) Handle(_ context.Context, request admission.Request) admission.Response {
-	requestLogger := logger.NewIDLogger(h.logger.WithValues(
-		"user", request.UserInfo.Username,
-		"resource", request.Resource, "name", request.Name,
-	))
-	if request.Namespace != "" {
-		requestLogger = requestLogger.WithValues("namespace", request.Namespace)
+// Handle checks the resource sizes.
+func (h *Handler) Handle(_ context.Context, req admission.Request) admission.Response {
+	var err error
+	switch req.Operation {
+	case admissionv1.Create:
+		err = h.handle(req)
+	case admissionv1.Update:
+		err = h.handle(req)
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("unknown operation request %q", req.Operation))
 	}
 
-	response := h.admitRequestSize(request, requestLogger)
-	if !response.Allowed {
-		metrics.RejectedResources.WithLabelValues(
-			fmt.Sprint(request.Operation),
-			request.Kind.Kind,
-			request.Namespace,
-			metricReasonSizeExceeded,
-		).Inc()
+	if err != nil {
+		var apiStatus apierrors.APIStatus
+		if errors.As(err, &apiStatus) {
```

Comment:
Can you explain when this type of error is returned. IIUC, the `handle` function only returns custom `fmt.Errorf` errors. 👀

Reply:
After https://github.com/gardener/gardener/pull/6873/commits/4a1f4ec696e9811f3aa1efcd02feda968d7002c1, it can also return `APIStatus` errors as requested by https://github.com/gardener/gardener/pull/6873#discussion_r1007786239.